zowe/zowe-common-c
# Module registry

## Proposed changes


To address potential LPA shortages when multiple ZIS instances are running on the same LPAR, a new feature--the module registry--has been introduced.

The module registry is a set of data structures in common storage in which eligible modules are traced. An eligible module is a module marked with a special macro; the mark and the module names are then used to uniquely identify the module. Once ZIS and its plug-in modules are marked, they can be added to the registry and shared: if another ZIS instance needs to load an identical module to the LPA, it will reuse a previously registered (and loaded to the LPA) module instead. Sharing identical modules should dramatically decrease the LPA footprint.

This feature must not be used when an application wants to remove its modules from the LPA (e.g., ZIS does that in dev mode) because it will affect any other applications using the same modules.

This PR addresses Issue: #405


## Type of change
Please delete options that are not relevant.

## PR Checklist
Please delete options that are not relevant.
- [x] My code follows the style guidelines of this project (see: [Contributing guideline](https://github.com/zowe/zss/blob/v2.x/master/CONTRIBUTING.md))

## Testing
* Clone this repo and check out the source branch for this change;
* Go to `./tests` and run `build_modreg.sh`;
* Make sure that the load library with the test module is APF-authorized;
* Make sure a ZIS instance has been run since the latest IPL;
* Use the JCL from `./tests/modregtest.c` to run the test program;
* The job must end with RC 0.

## Further comments
